Implements cpu_kernel_multiple_outputs and torch.frexp

RockingJavaBean opened the pull request:
Close #51108
Related #38349

This PR implements the `cpu_kernel_multiple_outputs` to support returning multiple values in a CPU kernel.
```c++
auto iter = at::TensorIteratorConfig()
  .add_output(out1)
  .add_output(out2)
  .add_input(in1)
  .add_input(in2)
  .build();

at::native::cpu_kernel_multiple_outputs(iter,
  [=](float a, float b) -> std::tuple<float, float> {
    float add = a + b;
    float mul = a * b;
    return std::tuple<float, float>(add, mul);
  }
);
```

The `out1` will equal to `torch.add(in1, in2)`, while the result of `out2` will be `torch.mul(in1, in2)`.
It helps developers implement new torch functions that return two tensors more conveniently, such as NumPy-like functions [divmod](https://numpy.org/doc/1.18/reference/generated/numpy.divmod.html?highlight=divmod#numpy.divmod) and [frexp](https://numpy.org/doc/stable/reference/generated/numpy.frexp.html#numpy.frexp).

This PR adds `torch.frexp` function to exercise the new functionality provided by `cpu_kernel_multiple_outputs`.

RockingJavaBean committed:
implement cpu_kernel_multiple_outputs to support returning multiple values

mruberry commented:
We should add a function that exercises this new functionality to test it, too. What about Numpy's frexp? (https://numpy.org/doc/stable/reference/generated/numpy.frexp.html).

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean commented:
@mruberry Thanks for the kind suggestion, I will update this PR by adding `torch.frexp` using both `cpu_kernel_multiple_outputs` and `gpu_kernel_multiple_outputs`.

RockingJavaBean committed:
add torch.frexp using cpu_kernel_multiple_outputs and gpu_kernel_multiple_outputs.

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
fix

RockingJavaBean committed:
CI fix and add multiple outputs support for UnaryUfuncInfo

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
fix flake8

RockingJavaBean committed:
fix test_native_functions_yaml and try resolve ROCM issues

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
disable frexp on ROCM

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
skip CUDA tests for ROCM platform

RockingJavaBean commented:
@mruberry I added `torch.frexp` with the proposed `cpu_kernel_multiple_outputs` and the existing `gpu_kernel_multiple_outputs` functionalities.
The `torch.frexp` is implemented to support promoting integral inputs to floating types, it differs from NumPy's frexp regarding the returning dtypes of `exponent`.

```python
>>> x = np.arange(5)
>>> np.frexp(x)
(array([0.  , 0.5 , 0.5 , 0.75, 0.5 ]), array([0, 1, 2, 2, 3], dtype=int32))
>>> y = torch.arange(5)
>>> torch.frexp(y)
torch.return_types.frexp(
mantissa=tensor([0.0000, 0.5000, 0.5000, 0.7500, 0.5000]),
exponent=tensor([0., 1., 2., 2., 3.]))
```

Numpy returns exponent as a `int32` array, while `torch.frexp` promotes the type to `float`.

RockingJavaBean commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1637,6 +1637,23 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+                       # Reference: https://github.com/pytorch/pytorch/pull/51097
+                       # numpy returns exponents as integral arrays, while torch.frexp promotes integral inputs
+                       # to floating types
+                       SkipInfo('TestUnaryUfuncs', 'test_reference_numerics'),
```
The reason to skip this test is https://github.com/pytorch/pytorch/pull/51097#issuecomment-774948814

RockingJavaBean commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -654,6 +654,70 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
+  // gpu_kernel_multiple_outputs also has issues of handling non contiguous tensors
+  // Reference: https://github.com/pytorch/pytorch/pull/51097
+  if (self.device().type() == DeviceType::CUDA) {
+    config.add_input(self.contiguous());
```
During my local development, I found the `gpu_kernel_multiple_outputs` does not behave correctly for `non-contiguous` tensors.
Calls `self.contiguous` before adding input to `TensorIterator` so that corresponding TestUnaryUfuncs checks would pass for CUDA path.

ngimel replied:
This was due to compiler bug, and was fixed in #51827 (which also adds a test exercising non-contiguous tensors with gpu_kernel_multiple_outputs).

RockingJavaBean replied:
@ngimel Thank you so much for this information.
I rebased my local development branch with latest code, the test cases related to `non-contiguous` tensors pass without calling `self.contiguous()` manullay.

RockingJavaBean commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -654,6 +654,70 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
+  TORCH_CHECK(at::isFloatingType(self.scalar_type()) || self.device().is_cpu(),
+              "frexp is not implemented for ",
+              self.scalar_type(),
+              " on CUDA device.");
```
The `gpu_kernel_multiple_outputs` uses `multi_outputs_unroll` for unrolling, and it seems `multi_outputs_unroll` cannot handling `casting` as it is implemented by `LoadWithoutCast` and `StoreWithoutCast`.
https://github.com/pytorch/pytorch/blob/bce4c82f0d2e4d36b96e5e10314c3bae16ed705e/aten/src/ATen/native/cuda/MemoryAccess.cuh#L263
Integral inputs cannot be promoted to floating types for CUDA inputs, and this check may be removed once the `gpu_kernel_multiple_outputs` is enhanced with casting support.

RockingJavaBean commented:
The `test/test_unary_ufuncs.py` and `test/test_ops.py` are extended to support testing unary functions that return multiple outputs. The `torch.frexp` passes all the test checks on `CPU`, which may prove the correctness of `cpu_kernel_multiple_outputs`.

Meanwhile, the `gpu_kernel_multiple_outputs` has issues regarding type casting and non-contiguous tensors, could we fill a new issue for this and try to enhance `gpu_kernel_multiple_outputs` in a follow-up PR?

RockingJavaBean commented:
@mruberry Please kindly take a look.

Changes for this PR
- `torch.frexp` is implemented with `cpu_kernel_multiple_outputs` and `gpu_kernel_multiple_outputs`.
- `torch.frexp` supports only floating types for CUDA inputs due to the casting issue of `gpu_kernel_multiple_outputs`.
- `torch.frexp` is skipped on ROCM platform, as the frexp function is not provided by AMD yet. ([HIP-MATH](https://rocmdocs.amd.com/en/latest/ROCm_API_References/HIP-MATH.html), https://github.com/ROCm-Developer-Tools/HIP/issues/2169)
- The test/test_unary_ufuncs.py and test/test_ops.py are extended to support testing unary functions that return multiple outputs.

I would love to hear your feedback and guidance, and will update this PR accordingly.

heitorschueroff commented:
@RockingJavaBean thank you for the updated, @mruberry and I will review this PR tomorrow.

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

heitorschueroff commented on test/test_unary_ufuncs.py:
```diff
@@ -300,6 +304,16 @@ def test_reference_numerics(self, device, dtype, op):
 
     # Tests for testing (dis)contiguity consistency
 
+    def _multiple_output_check_helper(self,
```
This is a really cool helper function for testing but we should consider whether the added complexity is worthwhile. There's probably some other ways to write this but let's focus on the rest of the PR first.

heitorschueroff commented on test/test_unary_ufuncs.py:
```diff
@@ -1574,6 +1639,25 @@ def test_atanh_domain_float(self, device, dtype):
         self.assertEqual(torch.isinf(torch.atanh(sample)), inf_mask)
         self.assertEqual(torch.isinf(sample.atanh()), inf_mask)
 
+    @unittest.skipIf(TEST_ROCM, "skip on ROCM")
```
We have a ``@skipCUDAIfRocm`` decorator.

heitorschueroff commented on test/test_unary_ufuncs.py:
```diff
@@ -1574,6 +1639,25 @@ def test_atanh_domain_float(self, device, dtype):
 
+    @unittest.skipIf(TEST_ROCM, "skip on ROCM")
+    @dtypes(*torch.testing.get_all_dtypes(include_complex=False, include_bfloat16=False))
+    def test_frexp(self, device, dtype):
```
Do we actually need this test if we have the OpInfo with a NumPy reference that will cross-check the results?

RockingJavaBean replied:
Thanks for pointing this out, I would remove this redundant test in the next update.

heitorschueroff commented on tools/autograd/derivatives.yaml:
```diff
@@ -497,6 +497,9 @@
   self: grad
 
+- name: frexp.Tensor(Tensor self) -> (Tensor mantissa, Tensor exponent)
+  self: grad * 1 / exponent.exp2()
```
cc @albanD could you help check this derivative for the ``torch.frexp`` function.

albanD replied:
That looks good.
You can remove the `1` though and just use `grad / exponent.exp2()`. But otherwise should be good.

RockingJavaBean replied:
@albanD Thank you so much for the kind suggestion, I will update this derivative accordingly.

heitorschueroff commented on torch/_torch_docs.py:
```diff
@@ -3238,6 +3238,32 @@ def merge_dicts(*dicts):
+           r"""
+frexp(input, *, out=None) -> (Tensor mantissa, Tensor exponent)
+
+Decompose :attr:`input` into mantissa and twos exponent, elementwise.
```
Decompose -> Decomposes

heitorschueroff replied:
```
Decomposes :attr:`input` into mantissa and exponent tensors such that :math:\text{input} = \text{mantissa} \times 2^{\text{exponent}}
```

heitorschueroff commented on torch/_torch_docs.py:
```diff
@@ -3238,6 +3238,32 @@ def merge_dicts(*dicts):
+
+
+Args:
+    input (Tensor): the input tensor to be decomposed where ``input = mantissa * 2**exponent``
```
Move the equation explaining the input tensor into the first paragraph.

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1637,6 +1637,23 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+                   dtypesIfCPU=all_types_and(torch.bool, torch.half),
+                   dtypesIfCUDA=floating_types_and(torch.half),
+                   # skip testing torch.frexp as it is not supported by ROCM platform yet
+                   dtypesIfROCM=_dispatch_dtypes(()),
```
You can use decorators instead as in

```
decorators=[skipCUDAIfRocm])
```

heitorschueroff reviewed (commented):
@RockingJavaBean, Mike and I reviewed the PR together and it's going in the right direction. This PR adds a lot of great new functionality so we really appreciate this contribution.

One thing to consider is if we can make it so the mantissa tensors are returned in the same floating dtype as the input and the exponent tensors are always returned in int32. This would be compatible with NumPy. An easy way to do this, though probably not the best, is to perform all computations in double dtype and then just before returning the results downcast to the correct types.

We would also really appreciate if you could file some issues for the things you highlighted with casting and contiguous tensors on CUDA devices.

Let us know if you have any other questions and we look forward to seeing the next update to the PR.

RockingJavaBean commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4205,6 +4205,15 @@
 - func: norm.names_out(Tensor self, Scalar? p, Dimname[1] dim, bool keepdim=False, *, Tensor(a!) out) -> Tensor(a!)
   use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
 
+- func: frexp.Tensor(Tensor self) -> (Tensor mantissa, Tensor exponent)
```
The reason for using overloaded function names here is that `aten::frexp` is already registered by JIT.
https://github.com/pytorch/pytorch/blob/e8ab58bfc78d513a4655c3668bebacbd733651e0/torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp#L885-L888

RockingJavaBean commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -654,6 +654,70 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
+  return config
+    .cast_common_dtype_to_outputs(true)
+    .promote_inputs_to_common_dtype(true)
+    .promote_integer_inputs_to_float(true)
```
@heitorschueroff Thanks so much for the advice, I will remove the type promotion part out of this PR.

RockingJavaBean commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -654,6 +654,70 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
+}
+
+std::tuple<Tensor, Tensor> frexp(const Tensor& self) {
+  frexp_check(self);
```
Thanks for pointing this out, I would update `torch.frexp` to let it only support float and double dtypes within this PR, and improve `torch.frexp` to support other dtypes in a follow-up PR.

RockingJavaBean commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -127,6 +127,63 @@ basic_loop(char* C10_RESTRICT data[], const int64_t* strides_, int64_t i, int64_
+
+template <typename func_t>
+static inline void
+multiple_outputs_loop(char* C10_RESTRICT data[], const int64_t* strides_, int64_t i, int64_t n, func_t&& op) {
```
It can handle multiple inputs.
This PR also updates `aten/src/ATen/test/tensor_iterator_test.cpp` by adding a binary ops test case for testing the functionality of `cpu_kernel_multiple_outputs`.
```c++
auto iter = at::TensorIteratorConfig()                                                            \
  .add_output(out1)                                                                               \
  .add_output(out2)                                                                               \
  .add_input(in1)                                                                                 \
  .add_input(in2)                                                                                 \
  .build();                                                                                       \
at::native::cpu_kernel_multiple_outputs(iter, [=](ctype a, ctype b) -> std::tuple<ctype, ctype> { \
  ctype add = a + b;                                                                              \
  ctype mul = a * b;                                                                              \
  return std::tuple<ctype, ctype>(add, mul);                                                      \
}); 
```

RockingJavaBean commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -597,6 +597,21 @@ static void rsqrt_kernel(TensorIterator& iter) {
 
+static void frexp_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND(kHalf,
+    iter.dtype(),
```
As this PR adopts type promotion previously by setting TensorIterator's flag like `_promote_integer_inputs_to_float` and `_cast_common_dtype_to_outputs`, the `iter.dtype()` here is a floating dtype.

mruberry replied:
Please add a comment that iter.dtype() is the dtype of the mantissa output, a floating point dtype that must be the same as the input's dtype

RockingJavaBean commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -252,6 +252,26 @@ void kaiser_window_kernel_cuda(TensorIterator& iter, int64_t window_length, doub
+  TORCH_CHECK(false, "torch.frexp is not implemented on ROCM platform.");
+#else
+  AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::Half,
+    iter.common_dtype(),
```
Sorry for not being consistent with the CPU kernel, I would unify its usage in the next update.

RockingJavaBean commented:
@heitorschueroff @mruberry I'm really grateful for your thorough review and patient guidance. 

To answer the question of whether we could use `std::tuple<scalar_t, int>` or `thrust::tuple<scalar_t, int>` for CPU and CUDA kernels, I have done the following experiment in my local development environment.

**1. explicitly define the dtypes of `mantissa` and `exponent`, removing type promotion logic**
```cpp
std::tuple<Tensor, Tensor> frexp(const Tensor& self) {
  Tensor mantissa = at::empty_like(self);
  Tensor exponent = at::empty_like(self, self.options().dtype(at::kInt));
  at::frexp_out(mantissa, exponent, self);
  return std::tuple<Tensor, Tensor>(mantissa, exponent);
}
std::tuple<Tensor&, Tensor&> frexp_out(const Tensor& self,
                                       Tensor& mantissa, Tensor& exponent) {
  auto iter = TensorIteratorConfig()
    .add_output(mantissa)
    .add_output(exponent)
    .add_input(self)
    .check_all_same_dtype(false)
    .set_check_mem_overlap(true)
    .build();
  frexp_stub(iter.device_type(), iter);
  return std::tuple<Tensor&, Tensor&>(mantissa, exponent);
}
```
**2. update CPU and CUDA kernels**
```cpp
cpu_kernel_multiple_outputs(
  iter,
  [](scalar_t a) -> std::tuple<scalar_t, int> {
    int exponent;
    scalar_t mantissa = std::frexp(a, &exponent);
    return std::tuple<scalar_t, int>(mantissa, exponent);
  }
);
gpu_kernel_multiple_outputs(iter, [=] GPU_LAMBDA (scalar_t a) -> thrust::tuple<scalar_t, int> {
  int exponent;
  scalar_t mantissa = std::frexp(a, &exponent);
  return {mantissa, exponent};
});
```

**3. Results**

```python
>>> x = torch.arange(5.)
>>> torch.frexp(x)
torch.return_types.frexp(
mantissa=tensor([0.0000, 0.5000, 0.5000, 0.7500, 0.5000]),
exponent=tensor([0, 1, 2, 2, 3], dtype=torch.int32))
>>> x = torch.arange(5., device='cuda')
>>> torch.frexp(x)
torch.return_types.frexp(
mantissa=tensor([0.0000, 0.5000, 0.5000, 0.7500, 0.5000], device='cuda:0'),
exponent=tensor([0, 1, 2, 2, 3], device='cuda:0', dtype=torch.int32))
```
Both `cpu_kernel_multiple_outputs` and `gpu_kernel_multiple_outputs` return the correct results,

For `frexp_out`, `cpu_kernel_multiple_outputs` is correct while `gpu_kernel_multiple_outputs` would fail to set an int64 `exponent` tensor due to potential casting issues.
```python
>>> x = torch.arange(5.)
>>> out1 = torch.full((5,), 1.)
>>> out2 = torch.full((5,), 2)
>>> out2.dtype
torch.int64
>>> torch.frexp(x, out=(out1, out2))
torch.return_types.frexp_out(
mantissa=tensor([0.0000, 0.5000, 0.5000, 0.7500, 0.5000]),
exponent=tensor([0, 1, 2, 2, 3]))
>>> out2.dtype
torch.int64

>>> x = torch.arange(5., device='cuda')
>>> out1 = torch.full((5,), 1., device='cuda')
>>> out2 = torch.full((5,), 2, device='cuda')
>>> out2.dtype
torch.int64
>>> torch.frexp(x, out=(out1, out2))
torch.return_types.frexp_out(
mantissa=tensor([0.0000, 0.5000, 0.5000, 0.7500, 0.5000], device='cuda:0'),
exponent=tensor([4294967296, 8589934594,          3,          2,          2],
       device='cuda:0'))
```

I would update this first PR to only support floating inputs for `torch.frexp` and file a new issue regarding the casting issue of `gpu_kernel_multiple_outputs`.

RockingJavaBean commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -252,6 +252,26 @@ void kaiser_window_kernel_cuda(TensorIterator& iter, int64_t window_length, doub
+      gpu_kernel_multiple_outputs(iter, [=] GPU_LAMBDA (scalar_t a) -> thrust::tuple<scalar_t, scalar_t> {
+        int exponent;
+        scalar_t mantissa = std::frexp(a, &exponent);
+        return {mantissa, static_cast<T_ACC>(exponent)};
```
Please refer to https://github.com/pytorch/pytorch/pull/51097#issuecomment-778354460

RockingJavaBean commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -597,6 +597,21 @@ static void rsqrt_kernel(TensorIterator& iter) {
+        [](scalar_t a) -> std::tuple<scalar_t, scalar_t> {
+          int exponent;
+          scalar_t mantissa = std::frexp(a, &exponent);
+          return std::tuple<scalar_t, scalar_t>(mantissa, static_cast<scalar_t>(exponent));
```
Yes, I believe we can. 
Please refer to https://github.com/pytorch/pytorch/pull/51097#issuecomment-778354460.

ngimel commented:
Kernel with multiple outputs is a niche thing, and it's unlikely that we are going to fix dynamic casting issues for gpu_kernel_multiple_output as increased compile time and binary size are not worth it. However, to make it fail loudly instead of silently, you can add `needs_dynamic_casting` check into gpu_kernel_multiple_outputs, similarly to how it's checked for regular gpu kernel https://github.com/pytorch/pytorch/blob/fa0a049d4ec8bd45cc17432b8bf5a0922f5294d3/aten/src/ATen/native/cuda/CUDALoops.cuh#L153. Even without dynamic casting, as you've demonstrated, gpu_kernel_multiple_outputs can produce scalar_t -> (scalar_t, int32_t) results, and that's enough to support floating point inputs. int64_t `out` argument can either error out, or you can manually handle casting on the gpu side of frexp implementation.

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
fix flake8 issues

RockingJavaBean committed:
update torch.frexp to return mantissa, exponent as a tuple of (scalar_t, int)

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
fix Windows CI

RockingJavaBean commented:
I have updated this PR with the following changes.
- drop the type promotion part while only support floating dtypes as input in this first PR.
- return `mantissa` tensors as the same dtype of input, and `exponent` tensors as `torch.int32` tensors.
- update the docs and test cases.

Thanks to @ngimel's kind help, `torch.frexp` no longer needs to call the `contiguous` method for CUDA inputs as `gpu_kernel_multiple_outputs` now can handle non-contiguous input correctly.
I also chose to explicitly error out the `non-int32` exponent tensor in the out argument in this first PR.

Please kindly take a look.
@heitorschueroff @mruberry

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

heitorschueroff commented:
@RockingJavaBean It looks like the PR picked up some changes from multi_dot PR, could you rebase and Mike and I will review it again. Thanks!

RockingJavaBean commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1841,7 +1857,7 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
                # These call into check_alias_annotation which does not support TensorList inputs
-               SkipInfo('TestCommon', 'test_variant_consistency_jit', 
+               SkipInfo('TestCommon', 'test_variant_consistency_jit',
                         dtypes=[torch.float32, torch.int32]),
```
Trimming the extra whitespace after the `SkipInfo` here causes the merge commit e1698c5 to contain the code change from `multi_dot`.

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
update the UnaryUfuncInfo of torch.frexp

RockingJavaBean committed:
fix test_reference_numerics_hard for torch.half

RockingJavaBean commented:
Thank you so much for reviewing this PR, I have rebased this PR and resolved conflicts. @heitorschueroff @mruberry 

The latest update of this PR also fixed an issue of the newly introduced `generate_numeric_tensors_hard` function.

https://github.com/pytorch/pytorch/blob/7a178a8a523d4653a3a2fa10c573b71e7fab1b9a/test/test_unary_ufuncs.py#L55-L59

The `generate_numeric_tensors_hard` used the `_large_float_vals` to generate large float numbers.
However, for `torch.half` dtype, it would generate `infinity` tensors as `4988429.2` and `1e20` far exceeds the range of `torch.half` (min=-65504, max=65504).

I also found `torch.frexp` returns a different exponent compared to NumPy through the Windows Cuda 10.1 CircleCI check.
Below are the results reproduced using my local Windows laptop.
```python
>>> x = torch.tensor(float('inf'), device='cuda')
>>> torch.frexp(x)
torch.return_types.frexp(
mantissa=tensor(inf, device='cuda:0'),
exponent=tensor(0, device='cuda:0', dtype=torch.int32))
>>> np.frexp(x.cpu().numpy())
>>> (inf, -1)
```
Fixing the `generate_numeric_tensors_hard` function enable this PR to test `torch.frexp` against `torch.half` on CUDA.

mruberry commented:
Thanks @RockingJavaBean! I'm still catching up from being out on vacation, but I'll take a look ASAP!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3241,6 +3241,34 @@ def merge_dicts(*dicts):
+Decomposes :attr:`input` into mantissa and exponent tensors
+such that :math:`\text{input} = \text{mantissa} \times 2^{\text{exponent}}`.
+
+The range of mantissa is the open interval(-1, 1).
```
"interval(-1, 1)" -> "interval (-1, 1)."

RockingJavaBean replied:
Addressed.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3241,6 +3241,34 @@ def merge_dicts(*dicts):
+
+The range of mantissa is the open interval(-1, 1).
+
+.. note:: This function requires inputs to be floating-point dtypes.
```
Instead of this note, add a "Supports..." statement like torch.sub() has (https://pytorch.org/docs/master/generated/torch.sub.html#torch.sub):

"Supports float inputs."

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1859,6 +1859,26 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+                   dtypesIfCUDA=floating_types_and(torch.half),
+                   # skip testing torch.frexp as it is not supported by ROCm platform yet
+                   decorators=[skipCUDAIfRocm],
+                   test_inplace_grad=False,
```
Set supports_tensor_out to False and could you revert the changes to test_out.py

RockingJavaBean replied:
Thank you so much for the tip, the changes to `test_out.py` have been reverted in the latest PR.

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1859,6 +1859,26 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+                       SkipInfo('TestUnaryUfuncs', 'test_reference_numerics_normal'),
+                       SkipInfo('TestUnaryUfuncs', 'test_reference_numerics_hard'),
+                       SkipInfo('TestUnaryUfuncs', 'test_reference_numerics_extremal'),
+                       SkipInfo('TestUnaryUfuncs', 'test_out_arg_all_dtypes'),
```
Rather then skipping this test, could you fix the test to query whether the op supports out tensor as in https://github.com/pytorch/pytorch/blob/baed2cfe0110610274f64d217af587a4a5c8dedd/test/test_ops.py#L343.

RockingJavaBean replied:
Addressed.

heitorschueroff reviewed (commented):
Hi @RockingJavaBean , @mruberry and I took another look at your updates together. Overall, this PR is looking pretty good and close to ready. We really appreciate all the work you put into implementing frexp and also taking extra time to fix bugs in our test suite such as separating the float values.

There are still a few suggestion we have for your review:
- We think we can reuse test_reference_numerics with a tweak, and while we appreciate you demonstrating that the other test_unary_ufunc.py tests can be adapted to work with frexp(), we're a little worried that change is too big. For the final PR, we'd like to suggest skipping those tests and leaving them unmodified.
- For the test_frexp_out, add cases for incorrectly sized and noncontiguous inputs
- Add supports_tensor_out=False to the UnaryUfuncInfo and fix the test_out_arg... to correctly query for the metadata instead of skipping the test.

We look forward to your next updates!

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean committed:
updates according to code reviews

RockingJavaBean committed:
add comments according to review

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs

RockingJavaBean commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -654,6 +654,38 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
+  // torch.frexp is implemented for floating-point dtypes for now,
+  // should add support for integral dtypes in the future.
+  TORCH_CHECK(at::isFloatingType(self.scalar_type()),
+              "frexp only supports floating-point dtypes");
```
Thanks so much for the kind advice, the error message has been updated in the latest PR.

RockingJavaBean commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -654,6 +654,38 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
+              "frexp only supports floating-point dtypes");
+
+  TORCH_CHECK(mantissa.dtype() == self.dtype(),
+              "Expected mantissa to have dtype ", self.dtype(), " but got ", mantissa.dtype());
```
Addressed.

RockingJavaBean commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -654,6 +654,38 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
+  TORCH_CHECK(mantissa.dtype() == self.dtype(),
+              "Expected mantissa to have dtype ", self.dtype(), " but got ", mantissa.dtype());
+  TORCH_CHECK(exponent.dtype() == at::kInt,
+              "Expected exponent to have int dtype but got ", exponent.dtype());
```
Addressed.

RockingJavaBean commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -206,6 +263,24 @@ void cpu_kernel(TensorIteratorBase& iter, func_t&& op) {
 }
 
+template <typename func_t>
+void cpu_kernel_multiple_outputs(TensorIteratorBase& iter, func_t&& op) {
```
The comments for `cpu_kernel_multiple_outputs` related functions are added in b20c25a236090ba9cdf3c3fe6d8c8636f322e822.

RockingJavaBean commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -598,6 +598,21 @@ static void rsqrt_kernel(TensorIterator& iter) {
+      cpu_kernel_multiple_outputs(
+        iter,
+        [](scalar_t a) -> std::tuple<scalar_t, int> {
+          int exponent;
```
addressed.

RockingJavaBean commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -252,6 +252,25 @@ void kaiser_window_kernel_cuda(TensorIterator& iter, int64_t window_length, doub
+    iter.dtype(),
+    "frexp_cuda", [&]() {
+      gpu_kernel_multiple_outputs(iter, [=] GPU_LAMBDA (scalar_t a) -> thrust::tuple<scalar_t, int> {
+        int exponent;
```
addressed.

RockingJavaBean committed:
updates

RockingJavaBean commented on test/test_unary_ufuncs.py:
```diff
@@ -570,6 +609,93 @@ def test_digamma(self, device, dtype):
 
+    @skipCUDAIfRocm
+    @dtypes(*torch.testing.get_all_fp_dtypes(include_half=True, include_bfloat16=False))
+    def test_frexp_reference_numerics(self, device, dtype):
```
Thanks so much for the kind advice, the latest PR tweaks `test_reference_numerics` for multi-output tensors with Python's built-in function `zip`.

RockingJavaBean replied:
The latest PR drops the custom reference_numerics test,  and tweaks the `test_reference_numerics` by comparing each output respectively for torch functions that return multiple values.

The checks for the `exponent` dtype are added in the custom `test_frexp_out` method as below
```python
self.assertTrue(exponent.dtype == torch.int32)
self.assertTrue(torch_to_numpy_dtype_dict[exponent.dtype] == np_exponent.dtype)
```

RockingJavaBean commented on test/test_unary_ufuncs.py:
```diff
@@ -570,6 +609,93 @@ def test_digamma(self, device, dtype):
+    @skipCUDAIfRocm
+    @dtypes(*torch.testing.get_all_fp_dtypes(include_half=True, include_bfloat16=False))
+    def test_frexp_out(self, device, dtype):
+        input = make_tensor((50, 50), device, dtype, low=-9, high=9)
```
addressed.

RockingJavaBean commented on test/test_unary_ufuncs.py:
```diff
@@ -570,6 +609,93 @@ def test_digamma(self, device, dtype):
+
+    @skipCUDAIfRocm
+    @dtypes(*torch.testing.get_all_fp_dtypes(include_half=True, include_bfloat16=False))
+    def test_frexp_out(self, device, dtype):
```
Thanks so much for the advice, the latest PR adds these cases to `test_frexp_out`.

RockingJavaBean commented on torch/_torch_docs.py:
```diff
@@ -3241,6 +3241,34 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> x = torch.arange(9.)
```
Addressed.

RockingJavaBean committed:
improve test

RockingJavaBean commented:
I'm really grateful for the thorough review and invaluable suggestions throughout this PR.

This PR has been updated with the following changes:
- update docs according to the suggestions, and add comments to the `cpu_kernel_multiple_outputs` related methods in `Loops.h`
- revert the changes to `test_ops.py`, as well as the changes to existing tests in `test_unary_ufuncs.py` except for the tweak for `test_reference_numerics`.
- add test cases for incorrectly sized and noncontiguous inputs to `test_frexp_out`, explicitly check `exponent` dtype, and compare it with NumPy counterpart in the customize test method `test_frexp`.

Please kindly take a look. @heitorschueroff @mruberry

heitorschueroff commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -712,6 +714,11 @@ std::tuple<Tensor&, Tensor&> frexp_out(const Tensor& self,
   return std::tuple<Tensor&, Tensor&>(mantissa, exponent);
 }
 
+// alias for lgamma, implements special.gammanln equivalent to
```
Were these changes picked up from another PR?

RockingJavaBean replied:
These changes were caused by the code conflicts. At that time, `UnaryOps.cpp` encountered the code conflicts between `torch.frexp` and `torch.special_gammaln`, I merged the latest code from `master` branch and resolved the conflicts in the merge commit 54472cc1f26271c56ad0f23abf01d490212ac0a1.

The above changes wouldn't show up in the final changes of this PR, you may confirm this via the `Files Changed` tab of Github (https://github.com/pytorch/pytorch/pull/51097/files)

![image](https://user-images.githubusercontent.com/40605299/110806006-9d552e80-82bc-11eb-8afd-4bd35b398727.png)

The above diff shows only the changes of `torch.frexp` are made to `UnaryOps.cpp`.

heitorschueroff replied:
You are right, thank you for clarifying this.

heitorschueroff reviewed (approved):
@RockingJavaBean This last version looks ready, great work! Just one last change before I land this, it looks like the PR picked up some changes from another PR which I commented on, could you confirm this and fix it please, I'll land it then.

RockingJavaBean commented:
@heitorschueroff I'm truly thankful for your kind review.
The above changes of `torch.special_gammaln` belongs to a merge commit fixing code conflicts, and would not be in the final changes of this PR, please kindly help confirm this.

heitorschueroff commented:
> @heitorschueroff I'm truly thankful for your kind review.
> The above changes of `torch.special_gammaln` belongs to a merge commit fixing code conflicts, and would not be in the final changes of this PR, please kindly help confirm this.

You're correct. I'm landing the PR now, thank you for this great PR.

mruberry reviewed (approved):
Nice job, @RockingJavaBean! This is a very technically complicated PR. I appreciate your thoughtfulness on both the technical challenges and the test architecture, too.

And thanks @heitorschueroff for reviewing this!

RockingJavaBean commented:
It is my honor to contribute to the PyTorch project and it cannot be done without your generous help and guidance.
I'm really grateful to you. @mruberry @heitorschueroff

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into cpu_kernel_mutiple_outputs
